torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -154,57 +154,93 @@ class ProcessGroupNCCL : public ProcessGroup {
         : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
-          value_(value),
+          value_(std::move(value)),
           deviceIndex_(deviceIndex),
-          cudaEvent_(cudaEvent) {}
+          cudaEvents_(cudaEvents) {
+      TORCH_INTERNAL_ASSERT(
+          cudaEvents_->size() == 1,
+          "FutureNCCL only supports single-process single-device mode.");
+    }
 
-    // Gets current stream of device and synchronizes recorded streams
+    // This constructor is used by then callback, it skips setting the value at
+    // the beginning. Later, the value will be set using markCompleted with the
+    // return value of callback.
+    explicit FutureNCCL(
+        c10::DeviceIndex deviceIndex,
+        std::shared_ptr<std::vector<at::cuda::CUDAEvent>> cudaEvents)
+        : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
+          deviceIndex_(deviceIndex),
+          cudaEvents_(cudaEvents) {
+      TORCH_INTERNAL_ASSERT(
+          cudaEvents_->size() == 1,
+          "FutureNCCL only supports single-process single-device mode.");
+    }
+
+    // Gets the current stream of the device and synchronizes recorded streams
     // with that. It will return after synchronizing the correct GPU streams to
     // ensure we can have async CUDA execution and it does not wait for the
     // entire operation to complete on GPU.
     void wait() override {
-      auto currentStream = at::cuda::getCurrentCUDAStream(deviceIndex_);
-      (*cudaEvent_).block(currentStream);
+      if (error_) {
+        throw *error_;
+      }
+      auto stream = at::cuda::getCurrentCUDAStream(deviceIndex_);
+      (*cudaEvents_)[0].block(stream);
+    }
+
+    // If FutureNCCL was created by FutureNCCL::then, its value would be empty
+    // initially. FutureNCCL::then will later use this method to set its value
+    // to the return value of the callback.
+    void markCompleted(at::IValue value) override {
+      TORCH_INTERNAL_ASSERT(
+          value_.isNone(),
+          "Attempting to set value of a FutureNCCL which has a value."
+          "FutureNCCL's value was internally set to NCCL collective's "
+          "outputs or the return value of the callback.");
+      value_ = std::move(value);
     }
 
-    // FutureNCCL's value is NCCL collective's outputs or callback() output with
-    // callbacks being invoked inline by addCallbackWithStream(), so
-    // markCompleted is not needed.
-    void markCompleted(at::IValue /* unused */) override {
-      C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
+    void setError(std::string err) override {
+      error_ = FutureError(std::move(err));
     }
 
-    // Just returns FutureNCCL's value.
+    // Just returns FutureNCCL's value after wait returns.
     at::IValue value() override {
+      wait();
       return value_;
     }
 
     const at::IValue& constValue() override {
+      wait();
```

Comment:
Add `TORCH_CHECK` here asserting that we have a value (hasValue)

Reply:
@pritamdamania87 If FutureNCCL doesn't have a value, there will be an error from the then callback and wait() will throw an error. Do we still need an additional assert here?

Reply:
What if there is no `.then` callback? In that case, `wait()` wouldn't throw an error right?